---
title: "Class10_Structural Bioinformatics (Focus on new AlphaFold2)"
author: "Changcheng Li (PID: A69027828)"
format: pdf
---

The main repository of structural data is the PDB. Let's examine what it contains.

I downloaded composition stats from:

```{r}
#
```


```{r}
stats <- read.csv("Data Export Summary.csv", row.names = 1)
stats
```

>Q: Write a function to fix this non numeric table...

We can use the 'gsub()' function

```{r}
x <- c('11', '101', 2, '1,001')
as.numeric(gsub(",","", x))
```

```{r}
rm.comma <- function(x){
  as.numeric(gsub(",", "", x))
}

apply(stats, 2, rm.comma)
```

Will add the rownames from the original wee table

```{r}
#rownames(pdbstats) <- c("Hello", "Lets", "take", "a", "break", "now")
#pdbstats
```


>Q1: What percentage of structures in the PDB are solved by X-Ray and Electron Microscopy.

```{r}
#totals <- apply(pdbstats, 2, sum)
#totals/totals["Total"]
```

```{r}
sum(stats$X.ray) / sum(stats$Total)
```
It is 84.83 %

>Q2: What proportion of structures in the PDB are protein?

```{r}
stats['Protein (only)', 'Total'] / sum(stats$Total)
```
It is 86.67%

>Q3: Type HIV in the PDB website search box on the home page and determine how many HIV-1 protease structures are in the current PDB?

It is 211377

#The PDB format

```{r}
#less ~/Downloads/1hsg.pdb         ## (use ‘q’ to quit)
```

>Q2-3 Lets skip this

# Using Mol* to examine HIV-Pr

Here is a rubish pic of HIV-Pr that is not very useful yet

![](1HSG.png)

>Q4: Water molecules normally have 3 atoms. Why do we see just one atom per water molecule in this structure?

Because it only shows the oxygen atom in water, hydrogen atoms are not shown because it is beyond the resolution of this structure (2A).

>Q5: There is a critical “conserved” water molecule in the binding site. Can you identify this water molecule? What residue number does this water molecule have.

It is HOH (308)

>Q6: Generate and save a figure clearly showing the two distinct chains of HIV-protease along with the ligand. You might also consider showing the catalytic residues ASP 25 in each chain and the critical water (we recommend “Ball & Stick” for these side-chains). Add this figure to your Quarto document.

Discussion Topic: Can you think of a way in which indinavir, or even larger ligands and substrates, could enter the binding site?
And a nicer pic colored by secondary structure with catalytic active site ASP 25 shown in each chain with MK1 drug and all important water...

![A lovely image](1HSG_lovely.png)

#3. Introduction to Bio3D in R

```{r}
library(bio3d)
```

```{r}
pdb <- read.pdb("1hsg")
```

```{r}
pdb
```

>Q7: How many amino acid residues are there in this pdb object? 

There are 198 aa residues.

>Q8: Name one of the two non-protein residues? 

MK1

>Q9: How many protein chains are in this structure? 

There are two chains: A and B.

```{r}
attributes(pdb)
```

```{r}
head(pdb$atom)
```

#Predicting functional motions of a single structure

```{r}
adk <- read.pdb("6s36")
```

```{r}
adk
```

```{r}
# Perform flexiblity prediction
m <- nma(adk)
```

```{r}
plot(m)
```

```{r}
mktrj(m, file="adk_m7.pdb")
```

#4. Comparative structure analysis of Adenylate Kinase

```{r}
# Install packages in the R console NOT your Rmd/Quarto file

#install.packages("bio3d")
#install.packages("devtools")
#install.packages("BiocManager")

#BiocManager::install("msa")
#devtools::install_bitbucket("Grantlab/bio3d-view")
```

>Q10. Which of the packages above is found only on BioConductor and not CRAN? 

msa is found only on BioConductor

>Q11. Which of the above packages is not found on BioConductor or CRAN?: 

bio3d-view is not found on BioConductor or CRAN

>Q12. True or False? Functions from the devtools package can be used to install packages from GitHub and BitBucket? 

True

#Search and retrieve ADK structures

```{r}
library(bio3d)
aa <- get.seq("1ake_A")
```

```{r}
aa
```

>Q13. How many amino acids are in this sequence, i.e. how long is this sequence? 

There are 214 amino acids.

```{r}
# Blast or hmmer search 
#b <- blast.pdb(aa)
```

```{r}
# Plot a summary of search results
#hits <- plot(b)
```

```{r}
# List out some 'top hits'
#head(hits$pdb.id)
```

```{r}
hits <- NULL
hits$pdb.id <- c('1AKE_A','6S36_A','6RZE_A','3HPR_A','1E4V_A','5EJE_A','1E4Y_A','3X2S_A','6HAP_A','6HAM_A','4K46_A','3GMT_A','4PZL_A')
```

```{r}
# Download releated PDB files
files <- get.pdb(hits$pdb.id, path="pdbs", split=TRUE, gzip=TRUE)
```

```{r}
# Align releated PDBs
pdbs <- pdbaln(files, fit = TRUE, exefile="msa")
```

```{r}
# Vector containing PDB codes for figure axis
ids <- basename.pdb(pdbs$id)

# Draw schematic alignment
#plot(pdbs, labels=ids)
```

```{r}
library(bio3d.view)
library(rgl)

view.pdbs(pdbs)
```

#Annotate collected PDB structures
```{r}
anno <- pdb.annotate(ids)
unique(anno$source)
```

```{r}
anno
```

#Principal component analysis

```{r}
# Perform PCA
pc.xray <- pca(pdbs)
plot(pc.xray)
```

```{r}
# Calculate RMSD
rd <- rmsd(pdbs)

# Structure-based clustering
hc.rd <- hclust(dist(rd))
grps.rd <- cutree(hc.rd, k=3)

plot(pc.xray, 1:2, col="grey50", bg=grps.rd, pch=21, cex=1)
```

#5. Optional further visualization

```{r}
# Visualize first principal component
pc1 <- mktrj(pc.xray, pc=1, file="pc_1.pdb")
```

```{r}
view.xyz(pc1)
view.xyz(pc1, col=vec2color( rmsf(pc1) ))
```

```{r}
#Plotting results with ggplot2
library(ggplot2)
library(ggrepel)

df <- data.frame(PC1=pc.xray$z[,1], 
                 PC2=pc.xray$z[,2], 
                 col=as.factor(grps.rd),
                 ids=ids)

p <- ggplot(df) + 
  aes(PC1, PC2, col=col, label=ids) +
  geom_point(size=2) +
  geom_text_repel(max.overlaps = 20) +
  theme(legend.position = "none")
p
```

#6. Normal mode analysis [optional]

```{r}
# NMA of all structures
modes <- nma(pdbs)
```

```{r}
plot(modes, pdbs, col=grps.rd)
```

>Q14. What do you note about this plot? Are the black and colored lines similar or different? Where do you think they differ most and why?

These proteins are clustered into three groups shown in three colors, proteins within the same cluster have similar line patterns.
The black and colored lines are similar in the overall trend but are quite different in some segments(flexible region).
They differ predominantly in flexible regions predicted above, these regions may see great changes in conformation among different proteins. These regions are especially segments with a secondary structure (α-helix or β-sheet). It suggests that although there can be various differences in the sequence of those proteins, the segments corresponding to secondary structures are quite conserved, which may represent structures that are important for the conserved functions of these proteins.